Multiarch builds

goshlanguage opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Hey gang, 

Love the project. In fact, I'd love to use it on my `arm64` clusters. I've modified the makefile to switch to docker's buildx plugin that leverages [buildkit](https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images) to build multiplatform images.

I've built and published this image at (docker hub) `hartje/oauth2-proxy`, deployed it to my local cluster, and went through the auth flow as a confirmation.

Adding or extending the list to support other architectures should be somewhat trivial. Would the project consider enabling this change? It would help a lot of us `arm64` folks out. 

Appreciate any feedback.

## Description

<!--- Describe your changes in detail -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [ ] I have created a feature (non-master) branch for my PR.

squat commented on Makefile:
```diff
@@ -13,7 +13,7 @@ MINIMUM_SUPPORTED_GO_MAJOR_VERSION = 1
 GO_VERSION_VALIDATION_ERR_MSG = Golang version is not supported, please update to at least $(MINIMUM_SUPPORTED_GO_MAJOR_VERSION).$(MINIMUM_SUPPORTED_GO_MINOR_VERSION)
 
-DOCKER_BUILD := docker build --build-arg VERSION=${VERSION}
+DOCKER_BUILD := docker buildx build --platform linux/amd64,linux/arm64,linux/arm/v6 --build-arg VERSION=${VERSION}
```
This causes a funny interaction with the building of images that are explicitly tagged per architecture (https://github.com/oauth2-proxy/oauth2-proxy/blob/633734f93d2255da9861ed16819f0daea53b0f41/Makefile#L48-L56).

Maybe those lines would be better off being removed?

JoelSpeed replied:
Initially, we will want to have the images tagged for both the normal tags (ie release number and latest) as well as the arch specific tags, to allow people to migrate across. Can we do that from the same build at all?

How does the go build within the dockerfile understand which architecture it should be building for? IIRC we set the GOOS and GOARCH specifically don't we?

goshlanguage replied:
Indeed, I meant to ask about deletion of those other builds, but didn't want to lead with code removal.

We could retag latest tag to the matrix of supported tags if that sounds good?

I don't understand how buildkit makes it work, but I haven't had the need to update GOOS and GOARCH when building multiarch containers from go binaries using `docker buildx build`

goshlanguage replied:
Perhaps in the initial link, `TARGETOS` and `TARGETARCH` are implicitly set to a set of build flags, `GOARCH` and `GOOS` being in that set

goshlanguage replied:
It can be explicitly passed through into the `Dockerfile` like so if that's preferable:
https://www.docker.com/blog/containerize-your-go-developer-environment-part-1/

JoelSpeed replied:
What if instead, we set the build and push commands to use buildx, but then set the platform for each of the lines in the later parts of the dockerfile, eg
```
docker:
	$(DOCKER_BUILD) --platform linux/amd64,linux/arm64,linux/arm/v6 -f Dockerfile -t $(REGISTRY)/oauth2-proxy:latest .

docker-all: docker
	$(DOCKER_BUILD) --platfom linux/amd64 -f Dockerfile -t $(REGISTRY)/oauth2-proxy:latest-amd64 .
	$(DOCKER_BUILD) --platfom linux/arm64 -f Dockerfile -t $(REGISTRY)/oauth2-proxy:latest-arm64 .
```
Does this kind of thing work?

goshlanguage replied:
Sorry for the long poll. That sounds good to me, I'll work on an update

JoelSpeed commented on Makefile:
```diff
@@ -57,17 +58,7 @@ docker-all: docker
-	docker push $(REGISTRY)/oauth2-proxy:${VERSION}-arm64
-	docker push $(REGISTRY)/oauth2-proxy:latest-armv6
-	docker push $(REGISTRY)/oauth2-proxy:${VERSION}-armv6
+	docker buildx build --push --platform linux/amd64,linux/arm64,linux/arm/v6 -t $(REGISTRY)/oauth2-proxy:latest .
```
Where did we get to on keeping the old tag structure during the push for a period?

goshlanguage replied:
ah, right! Sorry about that

JoelSpeed replied:
As far as I can tell, this still isn't resolved? Does that match your understanding?

goshlanguage replied:
It might be confusing if you're looking here only, if you look at the file review, the `docker-push-all` target looks like main. 
Am I missing something?

main:

```
.PHONY: docker-push-all
docker-push-all: docker-push
	docker push $(REGISTRY)/oauth2-proxy:latest-amd64
	docker push $(REGISTRY)/oauth2-proxy:${VERSION}
	docker push $(REGISTRY)/oauth2-proxy:${VERSION}-amd64
	docker push $(REGISTRY)/oauth2-proxy:latest-arm64
	docker push $(REGISTRY)/oauth2-proxy:${VERSION}-arm64
	docker push $(REGISTRY)/oauth2-proxy:latest-armv6
	docker push $(REGISTRY)/oauth2-proxy:${VERSION}-armv6
```

This branch:

```
.PHONY: docker-push-all
docker-push-all: docker-push
	docker push $(REGISTRY)/oauth2-proxy:latest-amd64
	docker push $(REGISTRY)/oauth2-proxy:${VERSION}
	docker push $(REGISTRY)/oauth2-proxy:${VERSION}-amd64
	docker push $(REGISTRY)/oauth2-proxy:latest-arm64
	docker push $(REGISTRY)/oauth2-proxy:${VERSION}-arm64
	docker push $(REGISTRY)/oauth2-proxy:latest-armv6
	docker push $(REGISTRY)/oauth2-proxy:${VERSION}-armv6
```
https://github.com/oauth2-proxy/oauth2-proxy/blob/8b085cd48050d8e6855cb62f4b10d11748acadf2/Makefile#L63-L71

JoelSpeed commented:
I think this is good to go, would you be able to add a changelog entry and then I'm happy to merge this

goshlanguage commented:
just add a changelog entry he says... 

For some reason, Github wouldn't accept my changelog entry. 

Ended up going through a back and forth trying to get it to clear, it kept saying the addition to changelog was a merge conflict. Tried to rebase to before the push, and ended up making a pretty good mess. 

I ended up dropping the commits I accidentally `fetch`ed in from upstream master since it was adding 51 commits to the PR, which wasn't as clean as I'd hoped. 

Lastly I pulled upstream master, added the changelog entry, and updated the Makefile to match the commit it was blessed from. 

Hope this isn't too cumbersome/difficult to follow.

JoelSpeed reviewed (approved):
Eeek, that doesn't sound so fun, code looks good though, thanks for all your work here!
